node/src/block_producer/block_producer_reducer.rs

```diff
@@ -12,68 +12,104 @@ use mina_p2p_messages::{
         StagedLedgerDiffBodyStableV1, StateBodyHash, StateHash, UnsignedExtendedUInt32StableV1,
     },
 };
-use openmina_core::constants::constraint_constants;
+use openmina_core::{block::ArcBlockWithHash, constants::constraint_constants};
 use openmina_core::{
-    block::AppliedBlock,
+    bug_condition,
     consensus::{
         global_sub_window, grace_period_end, in_same_checkpoint_window, in_seed_update_range,
         relative_sub_window,
     },
 };
+use redux::{callback, Dispatcher, Timestamp};
+
+use crate::{
+    transition_frontier::sync::TransitionFrontierSyncAction, Action, BlockProducerEffectfulAction,
+    State, Substate, TransactionPoolAction,
+};
 
 use super::{
-    calc_epoch_seed, to_epoch_and_slot, BlockProducerAction, BlockProducerActionWithMetaRef,
-    BlockProducerCurrentState, BlockProducerEnabled, BlockProducerState, BlockWithoutProof,
+    calc_epoch_seed, next_epoch_first_slot, to_epoch_and_slot,
+    vrf_evaluator::{
+        BlockProducerVrfEvaluatorAction, BlockProducerVrfEvaluatorState, InterruptReason,
+    },
+    BlockProducerAction, BlockProducerActionWithMetaRef, BlockProducerCurrentState,
+    BlockProducerEnabled, BlockProducerState, BlockWithoutProof,
 };
 
 impl BlockProducerState {
-    pub fn reducer(
-        &mut self,
-        action: BlockProducerActionWithMetaRef<'_>,
-        best_chain: &[AppliedBlock],
-    ) {
-        self.with_mut((), move |state| state.reducer(action, best_chain))
+    pub fn reducer(state_context: Substate<State>, action: BlockProducerActionWithMetaRef<'_>) {
+        BlockProducerEnabled::reducer(state_context, action);
     }
 }
 
 impl BlockProducerEnabled {
-    pub fn reducer(
-        &mut self,
-        action: BlockProducerActionWithMetaRef<'_>,
-        best_chain: &[AppliedBlock],
-    ) {
+    pub fn reducer(mut state_context: Substate<State>, action: BlockProducerActionWithMetaRef<'_>) {
```

Comment:
Please add a comment here explaining that `State` is accessed here and why. It will make our lives easier in the future.